File: StanfordSoftware/Samples/StanfordCoreNLP.FSharp.Samples/SUTime.fs
#if INTERACTIVE
#I @"..\packages\IKVM.7.3.4830.0\lib"
#I @"..\packages\Stanford.NLP.CoreNLP.3.2.0.1\lib"
#r @"..\packages\NUnit.2.6.3\lib\nunit.framework.dll"
#r "stanford-corenlp-3.2.0.dll"
#r "IKVM.Runtime.dll"
#r "IKVM.OpenJDK.Core.dll"
#endif

module SUTimeTests

open NUnit.Framework
open java.util
open edu.stanford.nlp.tagger.maxent
open edu.stanford.nlp.ling
open edu.stanford.nlp.pipeline
open edu.stanford.nlp.time
open edu.stanford.nlp.util
open Config

/// http://nlp.stanford.edu/software/sutime.shtml#Usage
let [<Test>] ``SUTime Defaut Test : Three interesting dates`` () =
    let pipeline = AnnotationPipeline()
    pipeline.addAnnotator(PTBTokenizerAnnotator(false))
    pipeline.addAnnotator(WordsToSentencesAnnotator(false))

    let tagger = MaxentTagger(! @"pos-tagger\english-bidirectional\english-bidirectional-distsim.tagger")
    pipeline.addAnnotator(POSTaggerAnnotator(tagger))

    let sutimeRules = 
        [| ! @"sutime\defs.sutime.txt";
           ! @"sutime\english.holidays.sutime.txt";
           ! @"sutime\english.sutime.txt" |]
        |> String.concat ","
    let props = Properties()
    props.setProperty("sutime.rules", sutimeRules ) |> ignore
    props.setProperty("sutime.binders", "0") |> ignore
    pipeline.addAnnotator(TimeAnnotator("sutime", props))

    let text = "Three interesting dates are 18 Feb 1997, the 20th of july and 4 days from today."
    let annotation = Annotation(text)
    annotation.set(CoreAnnotations.DocDateAnnotation().getClass(), "2013-07-14") |> ignore
    pipeline.annotate(annotation)

    printfn "%O\n" (annotation.get(CoreAnnotations.TextAnnotation().getClass()))
    let timexAnnsAll = annotation.get(TimeAnnotations.TimexAnnotations().getClass()) :?> java.util.ArrayList
    for cm in timexAnnsAll |> Seq.cast<CoreMap> do
        let tokens = cm.get(CoreAnnotations.TokensAnnotation().getClass()) :?> java.util.List
        let first = tokens.get(0)
        let last = tokens.get(tokens.size() - 1)
        let time = cm.get(TimeExpression.Annotation().getClass()) :?> TimeExpression
        printfn "%A [from char offset '%A' to '%A'] --> %A"
            cm first last (time.getTemporal())

File: StanfordSoftware/Samples/StanfordCoreNLP.CSharp.Samples/SUTime.cs
using System;

using java.lang;
using java.util;
using edu.stanford.nlp.tagger.maxent;
using edu.stanford.nlp.ling;
using edu.stanford.nlp.pipeline;
using edu.stanford.nlp.time;
using edu.stanford.nlp.util;

using NUnit.Framework;

using String = System.String;

namespace StanfordCoreNLP.CSharp.Samples
{
    [TestFixture]
    public class SUTimeTests
    {
        /// http://nlp.stanford.edu/software/sutime.shtml#Usage
        [Test]
        public void SUTimeDefautTest()
        {
            var pipeline = new AnnotationPipeline();
            pipeline.addAnnotator(new PTBTokenizerAnnotator(false));
            pipeline.addAnnotator(new WordsToSentencesAnnotator(false));

            var tagger =
                new MaxentTagger(
                    Config.GetModel(@"pos-tagger\english-bidirectional\english-bidirectional-distsim.tagger"));
            pipeline.addAnnotator(new POSTaggerAnnotator(tagger));

            var sutimeRules = new[] {
                                      Config.GetModel(@"sutime\defs.sutime.txt"),
                                      Config.GetModel(@"sutime\english.holidays.sutime.txt"),
                                      Config.GetModel(@"sutime\english.sutime.txt")
                                  };

            var props = new Properties();
            props.setProperty("sutime.rules", String.Join(",", sutimeRules));
            props.setProperty("sutime.binders", "0");
            pipeline.addAnnotator(new TimeAnnotator("sutime", props));

            const string text = "Three interesting dates are 18 Feb 1997, the 20th of july and 4 days from today.";
            var annotation = new Annotation(text);
            annotation.set(new CoreAnnotations.DocDateAnnotation().getClass(), "2013-07-14");
            pipeline.annotate(annotation);

            Console.WriteLine(annotation.get(new CoreAnnotations.TextAnnotation().getClass())+"\n");
            var timexAnnsAll = (ArrayList)annotation.get(new TimeAnnotations.TimexAnnotations().getClass());
            foreach (CoreMap cm in timexAnnsAll)
            {
                var tokens = (java.util.List)cm.get(new CoreAnnotations.TokensAnnotation().getClass());
                var first = tokens.get(0);
                var last = tokens.get(tokens.size() - 1);
                var time = (TimeExpression)cm.get(new TimeExpression.Annotation().getClass());
                Console.WriteLine("{0} [from char offset '{1}' to '{2}'] --> {3}",
                    cm, first, last, (time.getTemporal()));
            }
        }
    }
}
